docs/src/DeveloperDocumentation/printing.md

```diff
@@ -0,0 +1,638 @@
+3-element Vector{TypeofX{T}}
+:oneline
+:oneline
+:oneline
+```
+
+- `:details`; no restrictions except:
+  - the output must make sense as a standalone without context to non-specialists
+  - the number of output lines must fit in the terminal
+  - the row length should fit the terminal (as it is for julia matrices);
+  - if the object is simple enough use only one line
+- `:oneline`;  the output must print in one line
+  - should make sense as a standalone without context
+  - Variable names/generators/relations should not be printed only their number.
+- `:supercompact`; no nested printing
+  - a single term or a symbol/letter mimicking mathematical notation
+  - should usually only depend on the type and not of the type parameters or of
+    the concrete instance - exceptions of this rule are possible e.g. for `GF(2)`
+  - no nested printing. In particular variable names and `base_ring` must not be displayed.
+    This ensures that `:oneline` and `:supercompact` stay compact even for complicated things.
+    If you want nested printing use `:oneline`.
+    See it as a user readable version of the main (mathematical) type.
+
+Every type must implement `:oneline` printing, and may optionally implement `:details` or `:supercompact`.
+The fallback is always `:oneline`.
+
+One-line printing is used for substructures, e.g. the base ring, domain/codomain
+or in arrays, being in one line is important for formatting.
+
+Finally `:supercompact` is to be used for information coming from substructures
+or subsubstructures e.g. the `base_ring` of the `base_ring` will be printed
+`:supercompactly.`
+
+
+## Examples
+For each example we fix a single parent and display its three print modes
+consecutively each in its own block
+`:details`
+`:oneline`
+`:supercompact`
+### Coefficient rings
+```
+Integer ring
+```
+```
+Integer ring
+```
+
+```
+ZZ
+```
+-----
+```
+Rational field
+```
+```
+Rational field
+```
+```
+QQ
+```
+----
+##### Prime field
+```
+Finite field of characteristic 2
+```
+```
+Finite field of characteristic 2
+```
+```
+GF(2)
+```
+##### Finite field
+```
+Finite field of degree 7 over GF(29)
+```
+```
+Finite field of degree 7 over GF(29)
+```
+```
+GF(29^7)
+```
+##### Integer mod rings
+```
+Integers modulo 125
+```
+```
+Integers modulo 125
+```
+```
+ZZ/(125)  # ?
+```
+#### Numberfields
+#### A simple number field
+```
+Number field of degree 10
+with defining polynomial x^10 - 3x^9 + 7x^8 - 11x^7 + 13x^6 - 12x^5 + 9x^4 - 5x^3 + 3x^2 - 2x + 1
+  over QQ
+```
+```
+Number field of degree 10 over QQ
+```
+```
+Number field
+```
+##### Fancy number field
+```
+Number field of degree 10
+with defining polynomial x^10 - 3x^9 + 7x^8 - 11x^7 + 13x^6 - 12x^5 + 9x^4 - 5x^3 + 3x^2 - 2x + 1
+  over
+  Number field of degree 4
+  with defining polynomial 9x^4 - 5x^3 + 3x^2 - 3x + 1
+    over
+    Number field of degree 2
+    with defining polynomial x^2+1
+    over
+      QQ
+```
+```
+Number field of degree 2 over Number field
+```
```

Comment:
Am I misunderstanding the example or shouldn't the outermost be printed instead of the innermost here. 
Then it would read '...of degree 10...' in :oneline.